File: lib/android/app/src/main/java/com/reactnativenavigation/views/element/animators/YAnimator.java
package com.reactnativenavigation.views.element.animators;

import android.animation.Animator;
import android.animation.ObjectAnimator;
import android.graphics.Point;
import android.view.View;

import com.reactnativenavigation.utils.ViewUtils;
import com.reactnativenavigation.views.element.Element;

public class YAnimator extends PropertyAnimatorCreator<View> {

    private final int dy;

    public YAnimator(Element from, Element to) {
        super(from, to);
        final Point fromXy = ViewUtils.getLocationOnScreen(from.getChild());
        final Point toXy = ViewUtils.getLocationOnScreen(to.getChild());
        dy = fromXy.y - toXy.y;
    }

    @Override
    public boolean shouldAnimateProperty(View fromChild, View toChild) {
        return dy != 0;
    }

    @Override
    public Animator create() {
        return ObjectAnimator.ofFloat(to.getChild(), View.TRANSLATION_Y, dy, 0);
    }
}

File: lib/android/app/src/main/java/com/reactnativenavigation/views/element/animators/YAnimator.kt
package com.reactnativenavigation.views.element.animators

import android.animation.Animator
import android.animation.ObjectAnimator
import android.view.View
import android.view.View.TRANSLATION_Y
import android.view.ViewGroup
import com.facebook.react.views.text.ReactTextView
import com.reactnativenavigation.parse.SharedElementTransitionOptions
import com.reactnativenavigation.utils.ViewUtils
import com.reactnativenavigation.utils.withDuration

import com.reactnativenavigation.utils.withInterpolator
import com.reactnativenavigation.utils.withStartDelay

class YAnimator(from: View, to: View) : PropertyAnimatorCreator<View>(from, to) {
    private val dy: Int

    init {
        val fromXy = ViewUtils.getLocationOnScreen(from)
        val toY = (to.layoutParams as ViewGroup.MarginLayoutParams).topMargin
        dy = fromXy.y - toY
    }

    override fun shouldAnimateProperty(fromChild: View, toChild: View) = dy != 0

    override fun excludedViews() = listOf(ReactTextView::class.java)

    override fun create(options: SharedElementTransitionOptions): Animator {
        to.translationY = dy.toFloat()
        to.pivotY = 0f
        return ObjectAnimator
                .ofFloat(to, TRANSLATION_Y, dy.toFloat(), 0f)
                .withDuration(options.getDuration())
                .withStartDelay(options.getStartDelay())
                .withInterpolator(options.interpolation.interpolator)
    }
}